From 82cfeb8a1f39a20a4c65a1ef1d8beb228b238840 Mon Sep 17 00:00:00 2001
From: Ivan Vecera <ivecera@redhat.com>
Date: Mon, 8 Sep 2014 16:13:30 -0400
Subject: [ethernet] be2net: clear promiscuous bits in adapter->flags while disabling promiscuous mode

Message-id: <1410192881-30702-8-git-send-email-ivecera@redhat.com>
Patchwork-id: 92401
O-Subject: [RHEL7.1 PATCH 07/78] be2net: clear promiscuous bits in adapter->flags while disabling promiscuous mode
Bugzilla: 1088575
RH-Acked-by: Stefan Assmann <sassmann@redhat.com>
RH-Acked-by: Nikolay Aleksandrov <nikolay@redhat.com>

BZ: #1088575

Upstream commit(s):
commit 7ad09458a5be9a0990457c1a198e702559ac25ca
Author: Somnath kotur <somnath.kotur@emulex.com>
Date:   Mon Mar 3 14:24:43 2014 +0530

    be2net: clear promiscuous bits in adapter->flags while disabling promiscuous mode

    We should clear promiscuous bits in adapter->flags while disabling promiscuous
    mode. Else we will not put interface back into VLAN promisc mode if the vlans
    already added exceeds the maximum limit.

    Signed-off-by: Kalesh AP <kalesh.purayil@emulex.com>
    Signed-off-by: Somnath Kotur <somnath.kotur@emulex.com>
    Signed-off-by: David S. Miller <davem@davemloft.net>

Signed-off-by: Ivan Vecera <ivecera@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/emulex/benet/be_main.c b/drivers/net/ethernet/emulex/benet/be_main.c
index 0a48de1..b817182 100644
--- a/drivers/net/ethernet/emulex/benet/be_main.c
+++ b/drivers/net/ethernet/emulex/benet/be_main.c
@@ -1157,6 +1157,14 @@ ret:
  return status;
 }
 
+static void be_clear_promisc(struct be_adapter *adapter)
+{
+ adapter->promiscuous = false;
+ adapter->flags &= ~BE_FLAGS_VLAN_PROMISC;
+
+ be_cmd_rx_filter(adapter, IFF_PROMISC, OFF);
+}
+
 static void be_set_rx_mode(struct net_device *netdev)
 {
  struct be_adapter *adapter = netdev_priv(netdev);
@@ -1170,9 +1178,7 @@ static void be_set_rx_mode(struct net_device *netdev)
 
  /* BE was previously in promiscuous mode; disable it */
  if (adapter->promiscuous) {
-  adapter->promiscuous = false;
-  be_cmd_rx_filter(adapter, IFF_PROMISC, OFF);
-
+  be_clear_promisc(adapter);
   if (adapter->vlans_added)
    be_vid_config(adapter);
  }
-- 
1.7.1